contrib/infiniband/keys.ic

```diff
@@ -58,20 +62,59 @@ uint32_t update_rkey_restart(uint32_t rkey, int pd_id)
 
 void update_rkey_send(struct ibv_send_wr * wr, int pd_id)
 {
-  if (is_restart) {
-    if (wr == NULL) {
-      return;
-    }
-    if (wr->opcode == IBV_WR_RDMA_WRITE || 
-        wr->opcode == IBV_WR_RDMA_WRITE_WITH_IMM || 
-        wr->opcode == IBV_WR_RDMA_READ) {
-      wr->wr.rdma.rkey = update_rkey_restart(wr->wr.rdma.rkey, pd_id);
-    }
-    
-    if (wr->opcode == IBV_WR_ATOMIC_CMP_AND_SWP ||
-        wr->opcode == IBV_WR_ATOMIC_FETCH_AND_ADD) {
-      wr->wr.atomic.rkey = update_rkey_restart(wr->wr.atomic.rkey, pd_id);
-    }
-    update_rkey_send(wr->next, pd_id);
+  if (wr == NULL) {
+    return;
+  }
+
+  if (wr->opcode == IBV_WR_RDMA_WRITE ||
+      wr->opcode == IBV_WR_RDMA_WRITE_WITH_IMM ||
+      wr->opcode == IBV_WR_RDMA_READ) {
+    wr->wr.rdma.rkey = update_rkey_restart(wr->wr.rdma.rkey, pd_id);
+  }
+
+  if (wr->opcode == IBV_WR_ATOMIC_CMP_AND_SWP ||
+      wr->opcode == IBV_WR_ATOMIC_FETCH_AND_ADD) {
+    wr->wr.atomic.rkey = update_rkey_restart(wr->wr.atomic.rkey, pd_id);
+  }
+  update_rkey_send(wr->next, pd_id);
+}
+
+void update_ud_send_restart(struct ibv_send_wr *wr) {
+  struct internal_ibv_ah *internal_ah;
+  uint32_t size;
+  ibv_ud_qp_id orig_id, curr_id;
+
+  if (wr == NULL) {
+    return;
   }
+
+  internal_ah = ibv_ah_to_internal(wr->wr.ud.ah);
+  orig_id.qpn = wr->wr.ud.remote_qpn;
+  orig_id.lid = internal_ah->attr.dlid;
+
+  dmtcp_send_query_to_coordinator("qp_info",
+                                  &orig_id, sizeof(orig_id),
+                                  &curr_id, &size);
+
+  if (size == sizeof(ibv_ud_qp_id)) {
+    // The qp is created before checkpoint, qp_num changes
+    wr->wr.ud.remote_qpn = curr_id.qpn;
+  }
```

Comment:
The "if" statement checks whether there is a corresponding key-value pair in the coordinator database. `curr_id` and `size` are the output of `dmtcp_send_query_to_coordinator()`, and the if statement is true when there is a value corresponding to `orig_id` in the coordinator database.